File: matiasmasso/Mat.NET/DTO/DTOWinMenu.cs
using MatHelperStd;
using Newtonsoft.Json;

using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOWinMenuItem : DTOBaseGuid
    {
        public DTOWinMenuItem parent { get; set; }
        public DTOLangText langText { get; set; }
        public int ord { get; set; }
        public string action { get; set; }
        public Cods cod { get; set; }
        public CustomTargets customTarget { get; set; }
        public DTOBaseGuid tag { get; set; }

        public MimeCods Mime { get; set; }
        [JsonIgnore]
        public Byte[] icon { get; set; }

        public List<DTOEmp> emps { get; set; }
        public List<DTORol> rols { get; set; }
        public List<DTOWinMenuItem> children { get; set; }

        public enum Cods
        {
            notSet,
            folder,
            item
        }

        public enum CustomTargets
        {
            none,
            bancs,
            staff,
            reps
        }

        public DTOWinMenuItem() : base()
        {
            children = new List<DTOWinMenuItem>();
            langText = new DTOLangText();
            rols = new List<DTORol>();
        }

        public DTOWinMenuItem(Guid oGuid) : base(oGuid)
        {
            children = new List<DTOWinMenuItem>();
            langText = new DTOLangText();
            rols = new List<DTORol>();
        }

        public static DTOWinMenuItem Factory(DTOWinMenuItem oParent)
        {
            DTOWinMenuItem retval = new DTOWinMenuItem();
            {
                var withBlock = retval;
                withBlock.parent = oParent;
                withBlock.emps = oParent.emps;
            }
            return retval;
        }

        public static void loadBancs(List<DTOBanc> items, ref DTOWinMenuItem oParent)
        {
            if (oParent.children.Count == 0)
            {
                foreach (var item in items)
                {
                    DTOWinMenuItem oMenuitem = new DTOWinMenuItem();
                    {
                        var withBlock = oMenuitem;
                        withBlock.icon = item.Logo;
                        withBlock.tag = item;
                        withBlock.langText = new DTOLangText(item.abrOrNom());
                        withBlock.cod = DTOWinMenuItem.Cods.item;
                        withBlock.customTarget = DTOWinMenuItem.CustomTargets.bancs;
                    }
                    oParent.children.Add(oMenuitem);
                }
            }
        }

        public static void loadReps(List<DTORep> oReps, ref DTOWinMenuItem oParent)
        {
            if (oParent.children.Count == 0)
            {
                foreach (DTORep item in oReps)
                {
                    DTOWinMenuItem oMenuitem = new DTOWinMenuItem();
                    {
                        var withBlock = oMenuitem;
                        withBlock.icon = item.Img48;
                        withBlock.tag = item;
                        withBlock.langText = new DTOLangText(item.NickName);
                        withBlock.cod = DTOWinMenuItem.Cods.item;
                        withBlock.customTarget = DTOWinMenuItem.CustomTargets.reps;
                    }
                    oParent.children.Add(oMenuitem);
                }
            }
        }

        public static void loadStaffs(List<DTOStaff> oStaffs, ref DTOWinMenuItem oParent)
        {
            if (oParent.children.Count == 0)
            {
                foreach (DTOStaff item in oStaffs)
                {
                    DTOWinMenuItem oMenuitem = new DTOWinMenuItem();
                    {
                        var withBlock = oMenuitem;
                        withBlock.icon = item.Logo;
                        withBlock.tag = item;
                        withBlock.langText = new DTOLangText(item.Abr);
                        withBlock.cod = DTOWinMenuItem.Cods.item;
                        withBlock.customTarget = DTOWinMenuItem.CustomTargets.staff;
                    }
                    oParent.children.Add(oMenuitem);
                }
            }
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOWinMenu.vb
Public Class DTOWinMenuItem
    Inherits DTOBaseGuid

    Property Parent As DTOWinMenuItem
    Property LangText As DTOLangText
    Property Ord As Integer
    Property Action As String
    Property Cod As Cods
    Property CustomTarget As CustomTargets
    Property Tag As DTOBaseGuid
    <JsonIgnore> Property Icon As Image

    Property Emps As List(Of DTOEmp)
    Property Rols As List(Of DTORol)
    Property Children As List(Of DTOWinMenuItem)

    Public Enum Cods
        NotSet
        Folder
        Item
    End Enum

    Public Enum CustomTargets
        None
        Bancs
        Staff
        Reps
        UserTasks
    End Enum

    Public Sub New()
        MyBase.New()
        _Children = New List(Of DTOWinMenuItem)
        _LangText = New DTOLangText
        _Rols = New List(Of DTORol)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _Children = New List(Of DTOWinMenuItem)
        _LangText = New DTOLangText
        _Rols = New List(Of DTORol)
    End Sub

    Shared Function Factory(oParent As DTOWinMenuItem) As DTOWinMenuItem
        Dim retval As New DTOWinMenuItem()
        With retval
            .Parent = oParent
            .Emps = oParent.Emps
        End With
        Return retval
    End Function

    Shared Sub loadBancs(items As List(Of DTOBanc), ByRef oParent As DTOWinMenuItem)
        If oParent.Children.Count = 0 Then
            For Each item In items
                Dim oMenuitem As New DTOWinMenuItem
                With oMenuitem
                    .Icon = item.Logo
                    .Tag = item
                    .LangText = New DTOLangText(item.AbrOrNom)
                    .Cod = DTOWinMenuItem.Cods.Item
                    .CustomTarget = DTOWinMenuItem.CustomTargets.Bancs
                End With
                oParent.Children.Add(oMenuitem)
            Next
        End If
    End Sub

    Shared Sub loadReps(oReps As List(Of DTORep), ByRef oParent As DTOWinMenuItem)
        If oParent.Children.Count = 0 Then
            For Each item As DTORep In oReps
                Dim oMenuitem As New DTOWinMenuItem
                With oMenuitem
                    .Icon = item.Img48
                    .Tag = item
                    .LangText = New DTOLangText(item.NickName)
                    .Cod = DTOWinMenuItem.Cods.Item
                    .CustomTarget = DTOWinMenuItem.CustomTargets.Reps
                End With
                oParent.Children.Add(oMenuitem)
            Next
        End If
    End Sub

    Shared Sub loadStaffs(oStaffs As List(Of DTOStaff), ByRef oParent As DTOWinMenuItem)
        If oParent.Children.Count = 0 Then
            For Each item As DTOStaff In oStaffs
                Dim oMenuitem As New DTOWinMenuItem
                With oMenuitem
                    .Icon = item.Logo
                    .Tag = item
                    .LangText = New DTOLangText(item.Abr)
                    .Cod = DTOWinMenuItem.Cods.Item
                    .CustomTarget = DTOWinMenuItem.CustomTargets.Staff
                End With
                oParent.Children.Add(oMenuitem)
            Next
        End If
    End Sub

End Class
